### PR TITLE
[CI] Add remaining Ubuntu 24.04 Dockerfile

### PR DESCRIPTION
**Pull request by sarnex**
This adds the remaining Dockerfiles we'll need for Ubuntu 24.04. 

There will be two more containers added to the workflow after this PR, I can't add them now because they require changes in this PR and CI doesn't use the files from precommit.

### CODE REVIEW

File: devops/containers/ubuntu2404_build.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,42 @@
+FROM nvidia/cuda:12.6.3-devel-ubuntu24.04
```
Comment (uditagarwal97) [inline devops/containers/ubuntu2404_build.Dockerfile:1]: The Ubuntu 22 docker container uses CUDA 12.1.0 (https://github.com/intel/llvm/blob/sycl/devops/containers/ubuntu2204_build.Dockerfile#L1). Are we good to use CUDA 12.6.3 with DPC++?
Similarly, Ubuntu 22 container uses ROCM 6.1.1 while Ubuntu 24 uses ROCM 6.3.

**Review by uditagarwal97 — APPROVED** [review]
LGTM overall, just one question.

File: sycl/doc/developer/DockerBKMs.md [hunk 2]
```diff
@@ -59,9 +59,13 @@ The following containers are publicly available for DPC++ compiler development:
...
-   NVidia/AMD and can be used for building DPC++ compiler from source with all
-   backends enabled or for end-to-end testing with HIP/CUDA on machines with
-   corresponding GPUs available.
+   NVidia/AMD, is based on Ubuntu 22.04, and can be used for building DPC++
+   compiler from source with all backends enabled or for end-to-end testing
+   with HIP/CUDA on machines with corresponding GPUs available.
+- `ghcr.io/intel/llvm/ubuntu2404_build`: has development kits installed for
+   NVidia/AMD, is based on Ubuntu 24.04, and can be used for building DPC++
+   compiler from source with all backends enabled or for end-to-end testing
+   with HIP/CUDA on machines with corresponding GPUs available.
```
Comment (bader) [inline sycl/doc/developer/DockerBKMs.md:68]: Please, create two sub-sections for Ubuntu 22.04 and Ubuntu 24.04 based images.

File: devops/containers/ubuntu2404_build.Dockerfile (hunk 1 above)
Comment (sarnex) [reply devops/containers/ubuntu2404_build.Dockerfile:1]: Good question. Based on the Codeplay [doc](https://developer.codeplay.com/products/oneapi/nvidia/2025.0.0/guides/get-started-guide-nvidia) (who owns the AMD/NVidia support), they say CUDA12+ works but isn't they version they are currently testing on:

image

And for ROCM:

image

Here it doesn't list 6.3 specifically but says a wide variety should work. Also one of the runners is already using an unsupported GPU so it's not like we are following this doc today.

So IMO based on the doc I think it's fine to try to use the newer API versions but we can revert if there's some problem.

Comment (uditagarwal97) [reply devops/containers/ubuntu2404_build.Dockerfile:1]: Seems reasonable to me. @npmiller @JackAKirk FYI.

File: sycl/doc/developer/DockerBKMs.md [hunk 3]
```diff
@@ -36,20 +36,30 @@ identical for Docker and Podman. Choose whatever is available on your system.
 
 The following containers are publicly available for DPC++ compiler development:
 
-- `ghcr.io/intel/llvm/ubuntu2204_base`: contains basic Ubuntu 22.04 environment
+### Ubuntu 22.04-based images
+- `ghcr.io/intel/llvm/ubuntu2204_base`: contains basic environment
```
Comment (bader) [inline sycl/doc/developer/DockerBKMs.md:40]:
```suggestion
### Ubuntu 22.04-based images

- `ghcr.io/intel/llvm/ubuntu2204_base`: contains basic environment
```

File: sycl/doc/developer/DockerBKMs.md [hunk 4]
```diff
@@ -36,20 +36,30 @@ identical for Docker and Podman. Choose whatever is available on your system.
...
    setup for building DPC++ compiler from source.
-- `ghcr.io/intel/llvm/ubuntu2404_base`: contains basic Ubuntu 24.04 environment
-   setup for building DPC++ compiler from source.
-- `ghcr.io/intel/llvm/ubuntu2204_intel_drivers`: contains everything from the
-   Ubuntu 22.04 base container + pre-installed Intel drivers.
+  - `ghcr.io/intel/llvm/ubuntu2204_intel_drivers`: contains everything from the
+   base container + pre-installed Intel drivers.
```
Comment (bader) [inline sycl/doc/developer/DockerBKMs.md:43]:
```suggestion
- `ghcr.io/intel/llvm/ubuntu2204_intel_drivers`: contains everything from the
  base container + pre-installed Intel drivers.
```

File: sycl/doc/developer/DockerBKMs.md [hunk 5]
```diff
@@ -36,20 +36,30 @@ identical for Docker and Podman. Choose whatever is available on your system.
...
+- `ghcr.io/intel/llvm/ubuntu2204_build`: has development kits installed for
+   NVidia/AMD and can be used for building DPC++
+   compiler from source with all backends enabled or for end-to-end testing
+   with HIP/CUDA on machines with corresponding GPUs available.
+  - `ghcr.io/intel/llvm/sycl_ubuntu2204_nightly`: contains the latest successfully
+   built nightly build of DPC++ compiler. The image comes in three flavors:
+   with pre-installed Intel drivers (`latest`), without them (`no-drivers`) and
+   with development kits installed (`build`).
+### Ubuntu 24.04-based images
+- `ghcr.io/intel/llvm/ubuntu2404_base`: contains basic environment
```
Comment (bader) [inline sycl/doc/developer/DockerBKMs.md:59]:
```suggestion
   with development kits installed (`build`).

### Ubuntu 24.04-based images

- `ghcr.io/intel/llvm/ubuntu2404_base`: contains basic environment
```

**Review by bader — APPROVED** [review]

File: devops/containers/ubuntu2404_build.Dockerfile (hunk 1 above)
Comment (sarnex) [reply devops/containers/ubuntu2404_build.Dockerfile:1]: I'm going to merge this for now. This PR doesn't make us use the new versions in testing yet, it just adds the Docker images. There will be a subsequent PR to move the CI over to use those new images. Happy to address any feedback from the Codeplay team.

**Pull request merged**